### ISSUE TITLE
Pylint unaware of dataclass_transform type behavior

### ISSUE THREAD

**Issue by Quexington** [pylint-dev/pylint#9280]
### Bug description

Here's a minimal example of using a metaclass that turns its children into dataclasses automatically.

```python
from __future__ import annotations

from dataclasses import dataclass, field

from typing import Any
from typing_extensions import dataclass_transform


@dataclass_transform()
class SomeMeta(type):
    def __init__(cls: SomeMeta, *args: Any) -> None:
        dataclass()(cls)
        super().__init__(*args)


class SomeClass(metaclass=SomeMeta):
    _foo: int = field(metadata="blah")
```

I'm using a metaclass because it happens to be the way I noticed it in my own code but my guess is that it would not extend to other wrappers, etc. either.
```


### Configuration

_No response_

### Command used

```shell
pylint temp.py
```

### Pylint output

```python
temp.py:17:16: E3701: Invalid usage of field(), it should be used within a dataclass or the make_dataclass() function. (invalid-field-call)
```

### Expected behavior

No error

### Pylint version

```shell
pylint 3.0.2
astroid 3.0.1
Python 3.10.0 (tags/v3.10.0:b494f59, Oct  4 2021, 19:00:18) [MSC v.1929 64 bit (AMD64)]
```

### OS / Environment

Windows 11 powershell 7

**Comment by DanielNoord** [comment]
Duplicate of https://github.com/pylint-dev/pylint/issues/7437.
